build_tools/cmake/iree_static_linker_test.cmake

```diff
@@ -0,0 +1,259 @@
+  set(_O_FILE_NAME "${_RULE_NAME}.o")
+  set(_H_FILE_NAME "${_RULE_NAME}.h")
+
+  # Set common iree-compile flags
+  set(_COMPILER_ARGS ${_RULE_COMPILER_FLAGS})
+  list(APPEND _COMPILER_ARGS "--iree-hal-target-backends=llvm-cpu")
+  list(APPEND _COMPILER_ARGS  "--iree-llvm-link-embedded=false")
+  list(APPEND _COMPILER_ARGS  "--iree-llvm-link-static")
+  list(APPEND _COMPILER_ARGS "-iree-llvm-static-library-output-path=${_O_FILE_NAME}")
+  if(_RULE_TARGET_CPU_FEATURES)
+    list(APPEND _COMPILER_ARGS "-iree-llvm-target-cpu-features=${_RULE_TARGET_CPU_FEATURES}")
+  endif()
+  list(APPEND _COMPILER_ARGS "${_SRC_PATH}")
+
+  if(_RULE_EMITC)
+    set(_C_FILE_NAME "${_RULE_NAME}_emitc.h")
+    list(APPEND _COMPILER_ARGS "-iree-mlir-to-vm-c-module")
+    list(APPEND _COMPILER_ARGS "-o")
+    list(APPEND _COMPILER_ARGS "${_C_FILE_NAME}")
+
+    # Custom command for iree-compile to generate static library and C module.
+    add_custom_command(
+      OUTPUT
+        ${_H_FILE_NAME}
+        ${_O_FILE_NAME}
+        ${_C_FILE_NAME}
+      COMMAND ${_COMPILER_TOOL} ${_COMPILER_ARGS}
+      DEPENDS ${_COMPILER_TOOL} ${_RULE_SRC}
+    )
+
+    set(_EMITC_LIB_NAME "${_NAME}_emitc")
+    add_library(${_EMITC_LIB_NAME}
+      STATIC
+      ${_C_FILE_NAME}
+    )
+    target_compile_definitions(${_EMITC_LIB_NAME} PUBLIC EMITC_IMPLEMENTATION)
+    SET_TARGET_PROPERTIES(
+      ${_EMITC_LIB_NAME}
+      PROPERTIES
+        LINKER_LANGUAGE C
+    )
+  else()  # bytecode module path
+    set(_VMFB_FILE_NAME ${_RULE_NAME}.vmfb)
+    list(APPEND _COMPILER_ARGS "-o")
+    list(APPEND _COMPILER_ARGS "${_VMFB_FILE_NAME}")
+
+    # Custom command for iree-compile to generate static library and VM module.
+    add_custom_command(
+      OUTPUT
+        ${_H_FILE_NAME}
+        ${_O_FILE_NAME}
+        ${_VMFB_FILE_NAME}
+      COMMAND ${_COMPILER_TOOL} ${_COMPILER_ARGS}
+      DEPENDS ${_COMPILER_TOOL} ${_RULE_SRC}
+    )
+
+    # Generate the embed data with the bytecode module.
+    set(_MODULE_C_NAME "${_RULE_NAME}_module_c")
+    set(_EMBED_H_FILE_NAME ${_MODULE_C_NAME}.h)
+    set(_EMBED_C_FILE_NAME ${_MODULE_C_NAME}.c)
+    iree_c_embed_data(
+      NAME
+        "${_MODULE_C_NAME}"
+      IDENTIFIER
+        "${_NAME}"
+      GENERATED_SRCS
+        "${_VMFB_FILE_NAME}"
+      C_FILE_OUTPUT
+        "${_EMBED_C_FILE_NAME}"
+      H_FILE_OUTPUT
+        "${_EMBED_H_FILE_NAME}"
+      FLATTEN
+      PUBLIC
+    )
+  endif(_RULE_EMITC)
+
+  set(_LIB_NAME "${_NAME}_lib")
+  add_library(${_LIB_NAME}
+    STATIC
+    ${_O_FILE_NAME}
+  )
+  SET_TARGET_PROPERTIES(
+    ${_LIB_NAME}
+    PROPERTIES
+    LINKER_LANGUAGE C
+  )
+  # Set alias for this static library to be used later in the function.
+  add_library(${_PACKAGE_NS}::${_RULE_NAME}_lib ALIAS ${_LIB_NAME})
+
+  if(NOT _RULE_INPUT_SHAPES OR
+     NOT _RULE_INPUT_TYPE OR
+     NOT _RULE_HAL_TYPE OR
+     NOT _RULE_MAIN_FUNCTION OR
+     NOT _RULE_STATIC_LIB_PREFIX)
+    return()
+  endif()
```

Comment:
Either
A) document this behavior above ("if any of these arguments are not specified, only compilation will be tested")
B) emit a warning/error here (ideally right after `cmake_parse_arguments`, to fail fast)

If CMake had a concept of structs, we could group these together so it wouldn't be possible to set some but not all, but oh well :/ (maybe @GMNGeoffrey has ideas for how we could blunt this edge a bit)

Reply:
Add comments before this return.